OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs

```diff
@@ -140,29 +145,44 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
 			}
 			else
 			{
-				var aircraftInfo = transport.TraitOrDefault<Aircraft>();
-				if (aircraftInfo != null)
+				var aircraft = transport.TraitOrDefault<Aircraft>();
+				if (aircraft != null)
 				{
-					if (!aircraftInfo.IsPlane)
+					if (!aircraft.IsPlane)
 					{
-						transport.QueueActivity(new Turn(transport, aircraftInfo.Info.InitialFacing));
+						Move(transport, lastPosition);
+						transport.QueueActivity(new Turn(transport, aircraft.Info.InitialFacing));
 						transport.QueueActivity(new HeliLand(transport, true));
 					}
 					else
 					{
-						transport.QueueActivity(new Land(transport, Target.FromCell(transport.World, entryPath.Last())));
+						// Distance required for descent.
+						var altitude = aircraft.Info.CruiseAltitude.Length;
+						var landDistance = altitude * 1024 / aircraft.Info.MaximumPitch.Tan();
+						var angle = WAngle.FromFacing(aircraft.Facing);
+						var horizontal = landDistance * angle.Sin() / (1024 * 1024);
+						var vertical = landDistance * angle.Cos() / (1024 * 1024);
+
+						var delta = transport.CenterPosition - transport.World.Map.CenterOfCell(lastPosition);
+						var sign = delta.LengthSquared >= landDistance * landDistance ? 1 : -1;
+						var positionBeforeLanding = lastPosition + sign * new CVec((int)horizontal, (int)vertical);
```

Comment:
Something must be wrong with this. It causes the noticeable drop right before touchdown. If I add 1024 to the `landDistance`, the landing is smooth as expected, but I'm not sure if that's actually the part where the error is, and I forgot too much of my highschool math to understand what's happening here.